ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/ExecutionKeyFactory.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License; you may not use this file
+ * except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.plan;
+
+import io.confluent.ksql.execution.builder.KsqlQueryBuilder;
+import io.confluent.ksql.execution.context.QueryContext;
+import io.confluent.ksql.schema.ksql.PhysicalSchema;
+import io.confluent.ksql.serde.FormatInfo;
+import io.confluent.ksql.serde.WindowInfo;
+import io.confluent.ksql.testing.EffectivelyImmutable;
+import java.util.Objects;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.streams.kstream.Windowed;
+
+/**
+ * The {@code ExecutionKeyFactory} is in charge of creating the keys
+ * and the Serdes for the keys of any particular execution step.
+ *
+ * @param <K> the type of the key, usually either {@code Struct}
+ *            or {@code Windowed<Struct>}
+ */
+@EffectivelyImmutable
```

Comment:
For my own edification, what's the purpose of marking this EffectivelyImmutable? Is it just as a clue to the reader, or is it necessary since KStreamHolder is marked as Immutable?

Reply:
The latter (also I just copied it over from `KeySerdeFactory`, GH couldn't keep track that it's actually the same file because I made so many changes